### PR TITLE
Bspline eval overflow read

### PR DESCRIPTION
**Pull request by PDoakORNL**
Working on fix for #1369 
Have some question of how UBsplines of non NATURAL boundary conditions are supposed to behave.  Test programs conveniently did not demonstrate that.  They now push on this but which behavior is correct?

Can't comment yet on whether this effects the final numeric output of the code.

### CODE REVIEW

**Comment by prckent** [comment]
http://einspline.sourceforge.net/UBinterface.shtml ?

**Comment by prckent** [comment]
The historical idea was that the callers would always respect the required limits of the input parameters (whatever they happen to be...). Based on your recent experience, adding explicit checks might be wise. It would certainly be safer. Perhaps at least an assert?

**Comment by PDoakORNL** [comment]
I'll have to explore the implications here a little further really only QMCHamiltonian seems to consume the single spline.  Reading the einspline documentation on sourceforge, the main issue is lack of formalism I don't think the valid interval for evaluation is really defined that code doesn't bounds check at all and assumes it is [grid.start,grid.end). We already had bounds checking added in eval_multi_UBspline_3d_d possibly by @ye-luo  in 2015. But that i feel gives surprising values for at least for PERIODIC and ANTIPERIODIC grids. We use PERIODIC multi_UBsplines pretty frequently.

**Comment by ye-luo** [comment]
MPC under QMCHamiltonian is the only one evaluate single 3D spline. The MPC method only support periodic.
Remember NATURAL/PERIODIC/ANTIPERIODIC used in the spline class is different from the simulation system periodicity.

For construction, there is some difference between NATURAL and PERIODIC/ANTIPERIODIC. Take 1D as example. If the length of the box is 2 and the resolution is 1, values on grid point 0 and 1 needs to be provided to initialize the single spline with PERIODIC/ANTIPERIODIC. In the case of NATURAL, values on grid point 0,1,2 needs to be give. This is the reason why PERIODIC/ANTIPERIODIC has +3 points in the spline coefficients but NATURAL has +2 points.

I actually don't think we use NATURAL anywhere. We always use PERIODIC/ANTIPERIODIC in splines orbitals.

Before calling the evaluation of splines. PrimLattice.toUnit() or some std::floor computation is called to bring the position back to the primitive cell. However for the directions with open BC, no action is taken.
For this reason when the simulation system has open BC in some directions. Coordinates are no more guaranteed inside the box [0,1). Because spline is only defined within a box, protection is added when calling the evaluation by restricting ipart. I just realized that the residual part also needs to be altered when it hits min or max. @PDoakORNL is this you what you find strange? In real simulation with open BC, large boxes are used and orbitals are zero at the boarder grid points and the result is not affected by this issue.

**Comment by PDoakORNL** [comment]
I'm closing this.  As far as how we currently do things #1373 probably coverts this problem.  But even in spline2 our use of splines is a mess of assumptions. Features we are not really going to use need to be eliminated and the assumptions fully documented or spline2 needs to up its game.

**Comment by prckent** [comment]
Since the safest option has negligible cost and also happens to be simple, I don't think there is a good reason not to do the safest thing here. i.e. Define and document the behavior and then enforce it via e.g. hard error if an out of bounds position is passed in, or apply the appropriate modulus(coordinate,periodicity). Both are safe. The current situation is not.

File: src/einspline/bspline_eval_d_std.cpp [hunk 1]
```diff
@@ -28,43 +29,129 @@ extern const double* restrict   Ad;
...
   double u = x*spline->x_grid.delta_inv;
-  double ipart, t;
+  double ipart;
   t = modf (u, &ipart);
-  int i = (int) ipart;
+  i = ipart;
+  check_ipart(ipart, &(spline->x_grid), &(spline->xBC));
+}
+
+void check_coeff_offset_UBspline_2d_d(const UBspline_2d_d * const restrict spline,
```
Comment (ye-luo) [inline src/einspline/bspline_eval_d_std.cpp]: Can we just call the 1D twice or triple times instead of making a separate function for each case?

Comment (PDoakORNL) [reply src/einspline/bspline_eval_d_std.cpp]: I believe not because of the x_grid y_grid z_grid stuff.

Comment (PDoakORNL) [reply src/einspline/bspline_eval_d_std.cpp]: this is a nice illustration of why using names instead of template parameters is a bad idea

Comment (ye-luo) [reply src/einspline/bspline_eval_d_std.cpp]: Passing arguments?

Comment (ye-luo) [reply src/einspline/bspline_eval_d_std.cpp]: A simple operation replicated 6 times is not a good thing.

Comment (PDoakORNL) [reply src/einspline/bspline_eval_d_std.cpp]: previously replicated in every eval I didn't invent this library.

Comment (ye-luo) [reply src/einspline/bspline_eval_d_std.cpp]: I'm not saying the old way is good or bad. It was simple and fine. Now there are more stuff operated on each direction. It is worth adding a function but not one for each dimension case with contents duplicated. I prefer one function and called for each direction in the eval function.

Comment (PDoakORNL) [reply src/einspline/bspline_eval_d_std.cpp]: feel free

**Comment by PDoakORNL** [comment]
test this please

File: src/einspline/tests/test_one.cpp [hunk 2]
```diff
@@ -25,15 +26,15 @@
...
 TEST_CASE("double_1d_natural", "[einspline]")
 {
+  // two point case
   Ugrid x_grid;
   x_grid.start = 1.0;
   x_grid.end = 10.0;
   x_grid.num = 2;
 
-  double data[2];
+  double data[3];
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: The test was originally written with just 2.  This just allows me to run the new tests.

File: src/einspline/multi_bspline_eval_d_std3.cpp [hunk 3]
```diff
@@ -40,7 +40,10 @@ eval_multi_UBspline_1d_d (const multi_UBspline_1d_d *spline,
   x -= spline->x_grid.start;
   double ux = x*spline->x_grid.delta_inv;
   double ipartx, tx;
-  tx = modf (ux, &ipartx);  int ix = std::min(std::max(0,(int) ipartx),spline->x_grid.num-1);
+  tx = modf (ux, &ipartx);
+  // This protects from overflow reads of coefs but seems wrong for PERIODIC boundary
+  // conditions
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/multi_bspline_eval_d_std3.cpp]: I believe you added this code.  Please add a comment that you think explains what you did.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/multi_bspline_eval_d_std3.cpp]: Your commit from 2017 says it was for open boundary conditions only.  But isn't x_grid.num-1 only valid for PERIOD/ANTIPERIODIC splines?

File: src/einspline/tests/test_one.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: I assume test code shows expected behavior and proper usage.  I elected not to go with the fixes I wrote but simply assert on bad inputs and remove them from the tests since they result in overflow reads that trip ASAN.  Since the QMCPACK lore is that never should a value outside of the BC be requested and that is clearly the way einspline as a whole was implemented. If we want to replace it with well designed C++ library that is something we can look for or do.

File: src/einspline/multi_bspline_eval_d_std3.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/einspline/multi_bspline_eval_d_std3.cpp]: An open boundary system with plane wave orbitals generated with PBC. For this reason, splines are still PBC but electrons may walk outside the box. I will update the comment here.

File: src/einspline/tests/test_one.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/einspline/tests/test_one.cpp]: I'm thinking of just removing the 2 point test.

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: If you believe its incorrect we should.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/einspline/tests/test_one.cpp]: I agree with you that the caller code should handle the protection.
For this test particularly, do you remember where data[2] was accessed in the code? Is it the place where splines coefs are solved?

File: src/einspline/multi_bspline_eval_d_std3.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/multi_bspline_eval_d_std3.cpp]: ok and outside the box the function value should just be the terminal value of the spline?

File: src/einspline/tests/test_one.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: I think the find_coefs_xxx routines

File: src/einspline/multi_bspline_eval_d_std3.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/einspline/multi_bspline_eval_d_std3.cpp]: The current solution is using the terminal value but a more solid fix will be not allowing the electrons to move out of the box at all at the upper level even the system has open BC.

File: src/einspline/tests/test_one.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: and from there the solve

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: And then in the solve.  By inspection it seems like the solver might actually work for two data points.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/einspline/tests/test_one.cpp]: So after you changed 10 to 9.9999999, does the test pass ASAN?

[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/einspline/tests/test_one.cpp]: yes
